### PR TITLE
(feat)tiles: add boundingBox to Tile3D class

### PR DESCRIPTION
**Pull request by Pessimistress**
Related: https://github.com/visgl/deck.gl/pull/7795
To support Tile3DLayer + TerrainExtension in deck.gl, I'm proposing a unified API between 2d and 3d tile headers:

`tile.boundingBox: [min: number[], max: number[]]`

#### Change list
- Add `Tile3D.boundingBox` member
- Docs
- Test

### CODE REVIEW

File: modules/3d-tiles/wip/tile-3d-header.md [hunk 1]
```diff
@@ -31,9 +31,9 @@ Get the bounding volume of the tile's contents. This defaults to the
 tile's bounding volume when the content's bounding volume is
 `undefined`.
 
-### boundingSphere : BoundingSphere
```
Comment (Pessimistress) [inline modules/3d-tiles/wip/tile-3d-header.md:34]: This does not seem to exist

File: modules/tiles/src/tileset/tile-3d.ts [hunk 2]
```diff
@@ -303,6 +305,17 @@ export class Tile3D {
     return this._screenSpaceError;
   }
 
+  /**
+   * Get bounding box in cartographic coordinates
+   * @returns [min, max] each in [longitude, latitude, altitude]
+   */
+  get boundingBox(): [min: number[], max: number[]] {
+    if (!this._boundingBox) {
+      this._boundingBox = getCartographicBounds(this.header.boundingVolume, this.boundingVolume);
```
Comment (Pessimistress) [inline modules/tiles/src/tileset/tile-3d.ts:314]: To minimize perf impact, bounding box is only calculated on demand

File: modules/tiles/src/tileset/helpers/bounding-volume.ts [hunk 3]
```diff
@@ -197,3 +232,84 @@ function createSphere(sphere, transform, result?) {
...
+}
+
+function emptyCartographicBounds(): [number[], number[]] {
+  return [
+    [Infinity, Infinity, Infinity],
+    [-Infinity, -Infinity, -Infinity]
+  ];
+}
+
+function addToCartographicBounds(cartesian: Vector3, out: [number[], number[]]) {
```
[the remark this answers is not in the document]
Comment (ibgreen) [reply modules/tiles/src/tileset/helpers/bounding-volume.ts:306]: It seems to me that the target as it is not optional should come first? Also `Readonly` will make things even clearer here.

```suggestion
function addToCartographicBounds(out: [number[], number[]], cartesian: Readonly<Vector3>) {
```

**Review by ibgreen — APPROVED** [review]
Tests of the new utils are both indirect and weak, obviously, but OK for now.

File: modules/tiles/src/tileset/helpers/bounding-volume.ts [hunk 4]
```diff
@@ -68,6 +69,40 @@ export function createBoundingVolume(boundingVolumeHeader, transform, result) {
...
 
+/**
+ * Calculate the cartographic bounding box the tile's bounding volume.
+ * @param {Object} boundingVolumeHeader The tile's bounding volume header.
+ * @param {BoundingVolume} boundingVolume The bounding volume.
+ * @returns [min, max] in [longitude, latitude, altitude]
+ */
+export function getCartographicBounds(
+  boundingVolumeHeader,
+  boundingVolume: OrientedBoundingBox | BoundingSphere
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/tiles/src/tileset/helpers/bounding-volume.ts:80]: BoundingVolume is not exported by `@math.gl/culling`

File: modules/tiles/src/tileset/helpers/bounding-volume.ts [hunk 5]
```diff
@@ -197,3 +232,84 @@ function createSphere(sphere, transform, result?) {
 
   return new BoundingSphere(center, radius);
 }
+
+function orientedBoundingBoxToCartographicBounds(
+  boundingVolume: OrientedBoundingBox
+): [number[], number[]] {
```
[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/tiles/src/tileset/helpers/bounding-volume.ts:238]: `number[][]` is a standard format used by deck. Not sure about the loaders code base.

File: modules/tiles/src/tileset/tile-3d.ts (hunk 2 above)
Comment (Pessimistress) [reply modules/tiles/src/tileset/tile-3d.ts:314]: These are not needed by 3d tiles (hence we never calculated them before). Culling and rendering are done in the cartesian space.

Bounding boxes are regularly used in 2d tiles. The current implementation of `Tile2DHeader` in deck.gl takes a different format (`{west, east, south, north}` in geospatial viewports and `{left, right, bottom, top}` in non-geo viewports) which causes a lot of pains in typing. I am planning to converge the two.

File: modules/tiles/src/tileset/tile-3d.ts [hunk 6]
```diff
@@ -303,6 +309,17 @@ export class Tile3D {
     return this._screenSpaceError;
   }
 
+  /**
+   * Get bounding box in cartographic coordinates
+   * @returns [min, max] each in [longitude, latitude, altitude]
+   */
+  get boundingBox(): CartographicBounds {
```
Comment (belom88) [inline modules/tiles/src/tileset/tile-3d.ts:316]: The naming is mixed up with "oriented bounding box". I would prefer `extent` here . I3S before 1.7 version had extent2d https://github.com/Esri/i3s-spec/blob/master/docs/1.7/3DSceneLayer.cmn.md . From 1.8 there is 'fullExtent' with z coordinate -  https://github.com/Esri/i3s-spec/blob/master/docs/1.8/3DSceneLayer.cmn.md

**Review by belom88 — COMMENTED** [review]
The naming is mixed up with "oriented bounding box". I would prefer `extent` here . I3S before 1.7 version had extent2d https://github.com/Esri/i3s-spec/blob/master/docs/1.7/3DSceneLayer.cmn.md . From 1.8 there is 'fullExtent' with z coordinate -  https://github.com/Esri/i3s-spec/blob/master/docs/1.8/3DSceneLayer.cmn.md

**Comment by belom88** [comment]
v3.3.3 has been published

**Pull request merged**